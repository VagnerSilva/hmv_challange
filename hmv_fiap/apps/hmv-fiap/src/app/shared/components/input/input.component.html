<mat-form-field
	[appearance]="appearance"
	[formGroup]="parentFormGroup.form"
	class="input-container"
	(focusin)="onFocusIn($event)"
	(focusout)="onFocusOut($event)"
>
	<mat-label *ngIf="labelName">{{ labelName }}</mat-label>
	<input
		[formControl]="$any(ngControl.control)"
		matInput
		[placeholder]="placeholder"
		[required]="required"
		[type]="type"
	/>
	<ng-content></ng-content>
	<mat-hint>{{ hint }}</mat-hint>
	<mat-error
		class="form-text text-danger"
		*ngFor="
			let errorMessage of $any(parentFormGroup.form.controls)[$any(ngControl.name)]['errorMessages']
		"
	>
		{{ errorMessage }}<br />
	</mat-error>
</mat-form-field>
