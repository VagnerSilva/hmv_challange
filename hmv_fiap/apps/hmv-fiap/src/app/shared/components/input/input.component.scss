.mat-form-field-flex > .mat-form-field-infix {
	padding: 0.47em 0px !important;
}
.mat-form-field-label-wrapper {
	top: -1.5em;
}

.mat-form-field-appearance-outline.mat-form-field-can-float.mat-form-field-should-float
	.mat-form-field-label {
	transform: translateY(-1.1em) scale(0.75);
	width: 133.33333%;
}

.mat-form-field-appearance-outline .mat-form-field-wrapper {
	margin: 0 0 0.25em 0 !important;
}
