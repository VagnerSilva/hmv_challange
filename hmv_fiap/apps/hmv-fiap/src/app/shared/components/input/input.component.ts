import {
	Attribute,
	ChangeDetectionStrategy,
	ChangeDetectorRef,
	Component,
	ElementRef,
	Optional,
	Self,
	ViewEncapsulation,
} from '@angular/core'
import { FormGroupDirective, NgControl, NgForm } from '@angular/forms'
import { ErrorStateMatcher } from '@angular/material/core'
import { MatFormFieldAppearance } from '@angular/material/form-field'
import { InputType } from '../types'
import { InputBaseComponent } from '../utils/input-base.component'

@Component({
	selector: 'hmv-input',
	templateUrl: './input.component.html',
	styleUrls: ['./input.component.scss'],
	encapsulation: ViewEncapsulation.None,
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class InputComponent extends InputBaseComponent<InputComponent> {
	constructor(
		@Attribute(`labelName`) public labelName: string,
		@Attribute('appearance') public appearance: MatFormFieldAppearance = 'fill',
		@Attribute('hint') public hint: string,
		@Attribute('type') public type: InputType,
		@Optional() @Self() public override readonly ngControl: NgControl,
		@Optional() public override readonly parentFormGroup: FormGroupDirective,
		@Optional() public override readonly parentForm: NgForm,
		public readonly elementRef: ElementRef,
		public override readonly defaultErrorStateMatcher: ErrorStateMatcher,
		public override cdr: ChangeDetectorRef,
	) {
		super(elementRef, parentFormGroup, parentForm, defaultErrorStateMatcher, ngControl, cdr)
	}
}
