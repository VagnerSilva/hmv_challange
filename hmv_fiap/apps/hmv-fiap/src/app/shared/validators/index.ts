export class Validator {
	static msg = new Validator()

	REQUIRED = 'Campo obrigatório'

	MIN(value: number): string {
		return `Campo deve ter no mínimo ${value} caracteres`
	}
}
