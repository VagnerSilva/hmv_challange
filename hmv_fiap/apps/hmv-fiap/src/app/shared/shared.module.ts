import { CommonModule } from '@angular/common'
import { NgModule } from '@angular/core'
import { ReactiveFormsModule } from '@angular/forms'
import { RxReactiveFormsModule } from '@rxweb/reactive-form-validators'
import { ButtonRadioComponent } from './components/button-radio/button-radio.component'
import { ButtonComponent } from './components/button/button.component'
import { InputComponent } from './components/input/input.component'
import { SelectComponent } from './components/select/select.component'
import { TextareaComponent } from './components/textarea/textarea.component'
import { MaterialModule } from './material/material.module'

const components = [
	ButtonComponent,
	InputComponent,
	SelectComponent,
	TextareaComponent,
	ButtonRadioComponent,
]

const modules = [MaterialModule, ReactiveFormsModule]

@NgModule({
	declarations: [...components],
	imports: [CommonModule, ...modules],
	exports: [...components, ...modules, RxReactiveFormsModule],
})
export class SharedModule {}
