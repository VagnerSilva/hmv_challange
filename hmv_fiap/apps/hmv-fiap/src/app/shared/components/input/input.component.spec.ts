/* eslint-disable @typescript-eslint/no-explicit-any */
import { Component } from '@angular/core'
import { fakeAsync } from '@angular/core/testing'
import { FormBuilder, FormGroup, NgForm, ValidationErrors } from '@angular/forms'
import { createHostFactory, SpectatorHost } from '@ngneat/spectator/jest'
import { RxwebValidators } from '@rxweb/reactive-form-validators'
import { SharedModule } from '@shared/shared.module'
import { InputComponent } from './input.component'

@Component({ template: '' })
class CustomHostComponent {
	formExample!: FormGroup

	constructor(private fb: FormBuilder) {
		this.formExample = this.fb.group({
			name: [
				'',
				[
					RxwebValidators.minLength({
						message: `number error`,
						value: 3,
					}),
				],
			],
		})
	}
	title = 'Custom HostComponent'
}

const cases = {
	simples: `
    <form [formGroup]="formExample">
      <hmv-input formControlName="name"></hmv-input>
    </form>
  `,
	placeholder: `
  <form [formGroup]="formExample">
   <hmv-input formControlName="name" placeholder="fullname" ></hmv-input>
  </form>
  `,
	rightIcon: `
  <form [formGroup]="formExample">
   <hmv-input formControlName="name" placeholder="fullname" >
    <mat-icon matPrefix class="my-icon">person</mat-icon>
   </hmv-input>
  </form>
  `,
	hint: `
  <form [formGroup]="formExample">
   <hmv-input formControlName="name" placeholder="fullname" hint="hint_text" >
    <mat-icon matSuffix class="my-icon">person</mat-icon>
   </hmv-input>
  </form>
  `,
}

function formManager(host: SpectatorHost<InputComponent, CustomHostComponent>): {
	value: () => any
	valid: () => boolean
	errors: () => ValidationErrors | null | undefined
	touched: () => boolean
} {
	const form = host.hostComponent.formExample
	return {
		value: () => form.value,
		valid: () => form.valid,
		errors: () => form.get(`name`)?.errors,
		touched: () => form.touched,
	}
}

describe('InputComponent', () => {
	let host: SpectatorHost<InputComponent, CustomHostComponent>
	const createHost = createHostFactory({
		component: InputComponent,
		host: CustomHostComponent,
		imports: [SharedModule],
		mocks: [NgForm],
	})

	it('should display the host component input', fakeAsync(async () => {
		host = createHost(cases.simples)
		const input = host.query<HTMLInputElement>('input') as HTMLInputElement

		expect(input).toExist()

		host.typeInElement('name', input)

		const form = formManager(host)
		expect(form.value().name).toBe('name')
		expect(form.valid).toBeTruthy()

		host.typeInElement('na', input)
		expect(form.valid()).not.toBeTruthy()
		expect(host.hostFixture).toMatchSnapshot()
	}))

	it('should display placeholder', fakeAsync(async () => {
		host = createHost(cases.placeholder)
		const input = host.query<HTMLInputElement>('input') as HTMLInputElement
		expect(input).toHaveAttribute('data-placeholder', 'fullname')
		expect(host.hostFixture).toMatchSnapshot()
	}))

	it('Shows right icon', fakeAsync(async () => {
		host = createHost(cases.rightIcon)
		const icon = host.query(`mat-icon`)
		expect(icon).toHaveAttribute('matPrefix', '')
		expect(host.hostFixture).toMatchSnapshot()
	}))

	it('shows hint text', fakeAsync(async () => {
		host = createHost(cases.hint)
		const icon = host.query(`mat-icon`)
		const hint = host.query('mat-hint')
		expect(icon).toHaveAttribute('matSuffix', '')
		expect(hint).toHaveExactText('hint_text')
		expect(host.hostFixture).toMatchSnapshot()
	}))

	it('Show error message', fakeAsync(async () => {
		host = createHost(cases.hint)
		const input = host.query<HTMLInputElement>('input') as HTMLInputElement
		expect(input).toExist()
		host.typeInElement('na', input)
		const form = formManager(host)
		expect(form.errors()).not.toBeNull()
	}))
})
