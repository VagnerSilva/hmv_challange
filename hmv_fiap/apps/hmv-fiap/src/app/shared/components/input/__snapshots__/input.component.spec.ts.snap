// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputComponent Shows right icon 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  fb={[Function FormBuilder]}
  formExample={[Function FormGroup]}
  title={[Function String]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    <hmv-input
      class="ng-untouched ng-pristine ng-valid"
      formcontrolname="name"
      id="input-2"
      placeholder="fullname"
    >
      <mat-form-field
        appearance="outline"
        class="mat-form-field input-container mat-primary mat-form-field-type-mat-input mat-form-field-appearance-outline mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            <div
              class="mat-form-field-outline"
            >
              <div
                class="mat-form-field-outline-start"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-gap"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-outline mat-form-field-outline-thick"
            >
              <div
                class="mat-form-field-outline-start"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-gap"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-describedby="mat-hint-2"
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
                data-placeholder="fullname"
                matinput=""
              />
              <mat-icon
                aria-hidden="true"
                class="mat-icon notranslate my-icon material-icons mat-icon-no-color"
                data-mat-icon-type="font"
                matprefix=""
                role="img"
              >
                person
              </mat-icon>
              <span
                class="mat-form-field-label-wrapper"
              >
                <label
                  class="mat-form-field-label mat-empty mat-form-field-empty"
                >
                  <mat-label>
                  </mat-label>
                </label>
              </span>
            </div>
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
          >
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              <mat-hint
                class="mat-hint"
              >
              </mat-hint>
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
    </hmv-input>
  </form>
</ng-component>
`;

exports[`InputComponent should display placeholder 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  fb={[Function FormBuilder]}
  formExample={[Function FormGroup]}
  title={[Function String]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    <hmv-input
      class="ng-untouched ng-pristine ng-valid"
      formcontrolname="name"
      id="input-1"
      placeholder="fullname"
    >
      <mat-form-field
        appearance="outline"
        class="mat-form-field input-container mat-primary mat-form-field-type-mat-input mat-form-field-appearance-outline mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            <div
              class="mat-form-field-outline"
            >
              <div
                class="mat-form-field-outline-start"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-gap"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-outline mat-form-field-outline-thick"
            >
              <div
                class="mat-form-field-outline-start"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-gap"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-describedby="mat-hint-1"
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
                data-placeholder="fullname"
                matinput=""
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                <label
                  class="mat-form-field-label mat-empty mat-form-field-empty"
                >
                  <mat-label>
                  </mat-label>
                </label>
              </span>
            </div>
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
          >
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              <mat-hint
                class="mat-hint"
              >
              </mat-hint>
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
    </hmv-input>
  </form>
</ng-component>
`;

exports[`InputComponent should display the host component input 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  fb={[Function FormBuilder]}
  formExample={[Function FormGroup]}
  title={[Function String]}
>
  <form
    class="ng-untouched ng-dirty ng-invalid"
    novalidate=""
  >
    <hmv-input
      class="ng-untouched floating ng-dirty ng-invalid"
      formcontrolname="name"
      id="input-0"
    >
      <mat-form-field
        appearance="outline"
        class="mat-form-field input-container mat-primary mat-form-field-type-mat-input mat-form-field-appearance-outline mat-form-field-can-float mat-form-field-has-label ng-untouched _mat-animation-noopable mat-form-field-should-float mat-focused ng-dirty ng-invalid"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            <div
              class="mat-form-field-outline"
            >
              <div
                class="mat-form-field-outline-start"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-gap"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-outline mat-form-field-outline-thick"
            >
              <div
                class="mat-form-field-outline-start"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-gap"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-describedby="mat-hint-0"
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control ng-untouched cdk-text-field-autofill-monitored ng-dirty ng-invalid"
                data-placeholder=""
                matinput=""
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                <label
                  class="mat-form-field-label"
                >
                  <mat-label>
                  </mat-label>
                </label>
              </span>
            </div>
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
          >
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              <mat-hint
                class="mat-hint"
              >
              </mat-hint>
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
    </hmv-input>
  </form>
</ng-component>
`;

exports[`InputComponent shows hint text 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
  fb={[Function FormBuilder]}
  formExample={[Function FormGroup]}
  title={[Function String]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    <hmv-input
      class="ng-untouched ng-pristine ng-valid"
      formcontrolname="name"
      hint="hint_text"
      id="input-3"
      placeholder="fullname"
    >
      <mat-form-field
        appearance="outline"
        class="mat-form-field input-container mat-primary mat-form-field-type-mat-input mat-form-field-appearance-outline mat-form-field-can-float mat-form-field-has-label mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            <div
              class="mat-form-field-outline"
            >
              <div
                class="mat-form-field-outline-start"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-gap"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-outline mat-form-field-outline-thick"
            >
              <div
                class="mat-form-field-outline-start"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-gap"
                style="width: 0px;"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-describedby="mat-hint-3"
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
                data-placeholder="fullname"
                matinput=""
              />
              <mat-icon
                aria-hidden="true"
                class="mat-icon notranslate my-icon material-icons mat-icon-no-color"
                data-mat-icon-type="font"
                matsuffix=""
                role="img"
              >
                person
              </mat-icon>
              <span
                class="mat-form-field-label-wrapper"
              >
                <label
                  class="mat-form-field-label mat-empty mat-form-field-empty"
                >
                  <mat-label>
                  </mat-label>
                </label>
              </span>
            </div>
          </div>
          <div
            class="mat-form-field-subscript-wrapper"
          >
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              <mat-hint
                class="mat-hint"
              >
                hint_text
              </mat-hint>
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
          </div>
        </div>
      </mat-form-field>
    </hmv-input>
  </form>
</ng-component>
`;
